4.14.0/4.14.1 - Long compiler options in CFLAGS breaks configure’s flexlink test

Recently I was (again) trying to build Ocaml from this repository using Cygwin. I’ve downloaded release 4.14.0 from https://github.com/ocaml/ocaml/archive/refs/tags/4.14.0.tar.gz and if I run ./configure with no options, it will run smoothly and allow me to build properly. 

However, running something like `./configure CFLAGS="--param=ssp-buffer-size=4"` (or, I believe, due to some scant testing, anything with a double hyphen) results in the following error. This error also persists when building from the 4.14 development branch. 
```
checking for flexdll sources... no
checking for flexlink... flexlink
checking whether flexlink works... no
configure: error: flexlink does not work
```

config.log reveals the error is because of the option:
```
configure:13065: checking whether flexlink works
configure:13072: gcc -c --param=ssp-buffer-size=4  conftest.c >&5
configure:13072: $? = 0
configure:13091: flexlink -chain cygwin64 -exe -o conftest.exe --param=ssp-buffer-size=4 -DCAML_NAME_SPACE    conftest.c conftest2.o >&5
/usr/lib/flexdll/flexlink: unknown option '--param=ssp-buffer-size=4'.
```

It seems like the C compiler options shouldn’t be passed to flexlink—there’s a comment in configure that sort of describes this.
```
# Command to build executalbes
# In general this command is supposed to use the CFLAGs- and LDFLAGS-
# related variables (OC_CFLAGS and OC_LDFLAGS for ocaml-specific
# flags, CFLAGS and LDFLAGS for generic flags chosen by the user), but
# at the moment they are not taken into account on Windows, because
# flexlink, which is used to build executables on this platform, can
# not handle them.
```

I’ve discovered a temporary workaround but I’m not sure if it’s the "proper" one, so to speak. Some other environment variables are changed specifically for the flexlink test so they are saved before the test and reset to their saved values once the test is over. However, CFLAGS remains unchanged throughout this. 
```
  saved_CC="$CC"
  saved_CFLAGS="$CFLAGS"
  saved_CPPFLAGS="$CPPFLAGS"
  saved_LIBS="$LIBS"
  saved_ac_ext="$ac_ext"
  saved_ac_compile="$ac_compile"
  # Move the content of confdefs.h to another file so it does not
  # get included
  mv confdefs.h confdefs.h.bak
  touch confdefs.h
----------(the test)----------
  # Restore the content of confdefs.h
  mv confdefs.h.bak confdefs.h
  ac_compile="$saved_ac_compile"
  ac_ext="$saved_ac_ext"
  CPPFLAGS="$saved_CPPFLAGS"
  CFLAGS="$saved_CFLAGS"
  CC="$saved_CC"
  LIBS="$saved_LIBS"
```
so I’ve currently worked around this by clearing the CFLAGS variable between the save and the test, which shouldn’t be harmful since it gets restored in a couple lines or so, and after this change everything builds smoothly.
```
     CC="$flexlink -chain $flexdll_chain -exe"
     LIBS="conftest2.$ac_objext"
     CPPFLAGS="$internal_cppflags $CPPFLAGS"
+    CFLAGS=""
     cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 int main() { return 0; }
```

--- a/Changes
+++ b/Changes
@@ -482,6 +482,10 @@ OCaml 5.0
   removing it from the default namespace.
   (David Allsopp, review by Sébastien Hinderer)
 
+- #11370, #11373: Don't pass CFLAGS to flexlink during configure.
+  (David Allsopp, report by William Hu, review by Xavier Leroy and
+   Sébastien Hinderer)
+
 ### Bug fixes:
 
 - #10768, #11340: Fix typechecking regression when combining first class

--- a/aclocal.m4
+++ b/aclocal.m4
@@ -352,6 +352,7 @@ AC_DEFUN([OCAML_TEST_FLEXLINK], [
     CC="$1 -chain $2 -exe"
     LIBS="conftest2.$ac_objext"
     CPPFLAGS="$3 $CPPFLAGS"
+    CFLAGS=""
     AC_LINK_IFELSE(
       [AC_LANG_SOURCE([int main() { return 0; }])],
       [AC_MSG_RESULT([yes])],

--- a/configure
+++ b/configure
@@ -13083,6 +13083,7 @@ esac
     CC="$flexlink -chain $flexdll_chain -exe"
     LIBS="conftest2.$ac_objext"
     CPPFLAGS="$internal_cppflags $CPPFLAGS"
+    CFLAGS=""
     cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 int main() { return 0; }

